File: src/11.binarySearch/21.668.go
package array_greedy

func FindKthNumber(m int, n int, k int) int {
	l := 1 * 1
	r := m * n

	for l < r {
		mid := l + (r-l)/2

		count := 0
		i := m - 1
		j := 0

		for i >= 0 && j < n {
			if (i+1)*(j+1) <= mid {
				count += i + 1
				j++
			} else {
				i--
			}
		}

		if count < k {
			l = mid + 1
		} else {
			r = mid
		}
	}

	return r
}

File: src/11.binarySearch/21.668.ts
export function findKthNumber(m: number, n: number, k: number): number {
    let l = 1 * 1;
    let r = m * n;

    while (l < r) {
        const mid = l + Math.floor((r - l) / 2);

        let count = 0;
        let i = m - 1;
        let j = 0;

        // 找到 k 的范围
        while (i >= 0 && j < n) {
            if ((i + 1) * (j + 1) <= mid) {
                count += i + 1;
                j++;
            } else {
                i--;
            }
        }

        if (count < k) {
            l = mid + 1;
        } else {
            r = mid;
        }
    }

    return r;
}
